### ISSUE TITLE
Question about bibtex information in heredocs or docstrings

### ISSUE THREAD

**Issue by cfe316** [mcmtroffaes/sphinxcontrib-bibtex#279]
Dear maintainers,

I had a question about the sort of workflows that are possible with this package. I'm writing a package that references many academic works. My ideal solution has a number of properties:

1. Bibliographic information (i.e. raw bibtex entries) are available in the same file as the python classes. This is to make it easier for readers of the code to see where a formula comes from, without opening up another file or needing to reference the docs.
If points 2 and 3 could also work, I'd be willing to fall back to having bibliographic information in a separate file, either local or global.

2. It would be possible to take bibliographic references as strings and assign them to python objects, i.e. 
```
class Equation():
    def __init__(self):
        self.cite = """One or more bibtex entries as strings"""   
        # or like this
        self.cite = CITATIONS["MyCitationKey1"]
        ...
```
This is to support a [certain tool](https://openmdao.org/newdocs/versions/latest/other_useful_docs/om_command.html?highlight=cite) that then extracts `cite` attributes from a program and emits a bibtex file.

3. Operability with Sphinx. The bibtex citations can be referenced in classes' docstrings (I'm using `autodoc`) and linked in the html documentation.

4. I'm okay with repeating the raw bibtex entries per file (having a `moduleA.bib` for `moduleA.py`, and `moduleB.bib` for `moduleB.py`, with some references the same), but I'd prefer not to have to embed them multiple times in one file (like directly embedding bibtex strings in the `self.cite` attribute above.) That would certainly work but would become quite repetitive.

I'm not quite sure what a solution might look like, especially to support 2 and 3 together.
* Does this project, or another, support retrieving bibtex entries *as raw strings* from their citation keys? 
* Does this project support searching for bibtex references within a python file? This would make it effectively a polyglot between python and bibtex. I imagine this would look something like
```
CITATIONS = {'openmdao_2019': """
@article{openmdao_2019,
    Author={Justin S. Gray and John T. Hwang and Joaquim R. R. A.
            Martins and Kenneth T. Moore and Bret A. Naylor},
    Year={2019},
    ...
    }
"""
}
```
...
I've looked through the past issues here but haven't seen anything quite like this.

Alternately, if you have a better idea of how this might all be tied together, particularly if you have examples of projects with good reference management, I'd be interested to hear or see them.

Best wishes,
Jacob Schwartz

**Comment by drammock** [comment]
I'm not a maintainer, just a user, but here is an example of something vaguely similar to what I think you're seeking to do.

We (MNE-Python) have the following setup:
- a [single BibTeX file with all references](https://github.com/mne-tools/mne-python/blob/main/doc/references.bib)
- in individual tutorials or docstrings, relevant references are cited with `:footcite:` ([example](https://github.com/mne-tools/mne-python/blob/fecf5411d28f47dcb4ff379976a1b4813edc45a9/mne/preprocessing/maxwell.py#L119)) and displayed with `:footbibliography:` ([example](https://github.com/mne-tools/mne-python/blob/fecf5411d28f47dcb4ff379976a1b4813edc45a9/mne/preprocessing/maxwell.py#L203)) (cf. the references section of [the rendered docstring](https://mne.tools/dev/generated/mne.preprocessing.maxwell_filter.html))
- a [general bibliography](https://mne.tools/dev/bibliography.html) where all BibTeX entries are rendered *regardless of whether they were cited anywhere in our documentation*
- we hacked together the ability to show a couple of raw BibTeX entries on [this page](https://mne.tools/dev/overview/cite.html) (click the grey box to show) by inserting comments into our BibTeX file and making use of Sphinx's `.. include::` with options `:start-after:` and `:end-before:`. Here's [the source](https://github.com/mne-tools/mne-python/blob/main/doc/overview/cite.rst) for that page; see the corresponding comments [here](https://github.com/mne-tools/mne-python/blob/fecf5411d28f47dcb4ff379976a1b4813edc45a9/doc/references.bib#L6), [here](https://github.com/mne-tools/mne-python/blob/fecf5411d28f47dcb4ff379976a1b4813edc45a9/doc/references.bib#L16), and [here](https://github.com/mne-tools/mne-python/blob/fecf5411d28f47dcb4ff379976a1b4813edc45a9/doc/references.bib#L27)

This last bullet is potentially a solution to your point 3, though perhaps too ad-hoc to really work well for your use case, as it would probably require bracketing every BibTeX entry with something like `% begin-GramfortEtAl2013a` and `% end-GramfortEtAl2013a` and then manually adding the `.. include::` directive each time you `:cite:`, `:citep:`, or `:footcite:`.  But it might give you some ideas as to how to make it more automated?

**Comment by mcmtroffaes** [comment]
To be clear, this extension uses pybtex to render references in bibtex format (from one or more bib files) into restructured text, as part of the sphinx builder (or as part of the docutils builder, with much more limited functionality). That's all it aims to do. As for the specific points:

1. sphinxcontrib-bibtex assumes one or more global bib files. These are parsed at the very start of the sphinx build process (even before any docs are read), to have all information available to resolve and render citations later on in the process. Bibtex source code is not assumed to come from python sources. The part of sphinx that looks at python is the autodoc extension, and sphinxcontrib-bibtex works independently of autodoc (however see point 3).
2. I'm sorry but I don't quite understand this point. Either way, to bridge python and bibtex, I believe you're looking for pybtex (which sphinxcontrib-bibtex uses). I suggest to have a look there.
3. sphinxcontrib-bibtex fully supports citations in docstrings with autodoc. For an example (part of the regression testing), see https://github.com/mcmtroffaes/sphinxcontrib-bibtex/tree/develop/test/roots/test-autodoc Specifically, your docstrings can contain cite roles and bibliography directives, and these will be treated as restructured text sources as usual, by sphinx. The bibtex source still resides in bib files.
4. sphinxcontrib-bibtex (and pybtex) assume every bibtex entry is specified only once across the entire project as each key must point to a unique entry.
5. Perhaps pybtex does this (I'm not sure though, I've never tried).
6. Not beyond point 3.

**Comment by cfe316** [comment]
Thank you, @drammock and @mcmtroffaes . I've found a solution that I'm happy with, similar to @drammock 's. 

1. I gave up on having all the bibtex information directly within a source file; it if were in multiple places that would be unnecessary duplication anyway. I'll just have a single `refs.bib`.

For point 2 I can use pybtex to generate a string corresponding to a bibtex entry; that's just what I needed.

3. Footcite and footbibliography work great.

Thanks again!
